### PR TITLE
Introduce Microsoft.Terminal.UI and consolidate UI helpers

### PR DESCRIPTION
**Pull request by DHowett**
This pull request introduces the module Microsoft.Terminal.UI.dll, and moves into it the following things:

- Any `IDirectKeyListener`
- All XAML converter helpers from around the project
   - ... including `IconPathConverter` from TerminalSettingsModel
   - ... but not `EmptyStringVisibilityConverter`, which has died

It also adds a XAML Markup Extension named `mtu:ResourceString`, which will allow us to refer to string resources directly from XAML. It will allow us to remove all of the places in the code where we manually set resources on XAML controls.

### CODE REVIEW

File: src/cascadia/UIHelpers/IconPathConverter.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+#include "IconPathConverter.g.h"
+
+namespace winrt::Microsoft::Terminal::UI::implementation
+{
+    struct IconPathConverter
+    {
+        IconPathConverter() = default;
+
+        static Windows::UI::Xaml::Controls::IconElement IconWUX(const winrt::hstring& iconPath);
```
Comment (DHowett) [inline src/cascadia/UIHelpers/IconPathConverter.h:11]: there is a difference between IconWUX and IconSourceWUX in that one of them supports `foo.dll,3` and the other does not. Which one does is left as an exercise for the reader. Why it does that is left as an exercise for @PankajBhojwani and @zadjii-msft.

**Comment by DHowett** [comment]
msbuild vcxproj ProjectDependencies
is a tire fire

**Comment by DHowett** [comment]
discussion: what if we just disable the PCH for this one and _make our include stack better_

**Comment by DHowett** [comment]
At this point, I have forgotten why it is not ready for review. I'm sure I will figure out why soon enough.

File: src/cascadia/UIHelpers/ResourceString.cpp [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#include "pch.h"
+#include "ResourceString.h"
+#include "ResourceString.g.cpp"
+
+namespace winrt::Microsoft::Terminal::UI::implementation
+{
+    winrt::Windows::Foundation::IInspectable ResourceString::ProvideValue()
```
Comment (zadjii-msft) [inline src/cascadia/UIHelpers/ResourceString.cpp:7]: we're not actually using this anywhere yet, right? just a "we can if we need it" thing? Maybe that's why we didn't merge this before?

**Review by zadjii-msft — APPROVED** [review]
I'd reckon my earlier reluctance to merging this was because it's _another_ GB of pch we need to carry around. But also, meh. I'm sure it's fine. Just last week I needed some of those TSE converters in TerminalApp so :shipit:

**Review by lhecker — APPROVED** [review]

**Comment by lhecker** [comment]
> I'd reckon my earlier reluctance to merging this was because it's another GB of pch we need to carry around.

Nowadays there's a very nicely written doc for header units: https://learn.microsoft.com/en-us/cpp/build/walkthrough-header-units?view=msvc-170  I think we should migrate to that. In fact, they even say:
> Header units are the recommended alternative to precompiled header files (PCH).

**Comment by zadjii-msft** [comment]
> Nowadays there's a very nicely written doc for header units

I have a couple threads with Scott and Cameron about doing _just that_. There was just _s o m e t h i n g &nbsp; e l s e_ going on for the last 6 months that prevented me from caching that in for long enough 😂. I reckon I'll probably give that a shot this year

**Comment by lhecker** [comment]
I'm going to resolve the conflicts and merge this, because I want to use the `EmptyStringVisibilityConverter` in my PR.

**Review by lhecker — APPROVED** [review]

**Comment by DHowett** [comment]
I'm surprised the existing uses of `local:Converters` didn't merge conflict

**Comment by DHowett** [comment]
@zadjii-msft i will need a second tap on this due to the merge policy

**Review by zadjii-msft — APPROVED** [review]

**Comment by zadjii-msft** [comment]
oof this broke the canary builds pretty bad 

https://microsoft.visualstudio.com/Dart/_build/results?buildId=92144278&view=results

**Comment by DHowett** [comment]
> oof this broke the canary builds pretty bad
> 
> https://microsoft.visualstudio.com/Dart/_build/results?buildId=92144278&view=results

#16811

**Pull request merged**